pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,446 @@
+		return
+	}
+	c.queueEndpoints(obj)
+}
+
+// onEndpointsUpdate queues a sync for the relevant Endpoints resource.
+func (c *Controller) onEndpointsUpdate(prevObj, obj interface{}) {
+	endpoints := obj.(*v1.Endpoints)
+	prevEndpoints := prevObj.(*v1.Endpoints)
+	if endpoints == nil || prevEndpoints == nil {
+		utilruntime.HandleError(fmt.Errorf("onEndpointsUpdate() expected type v1.Endpoints, got %T, %T", prevObj, obj))
+		return
+	}
+	if !c.shouldMirror(endpoints) && !c.shouldMirror(prevEndpoints) {
+		klog.V(5).Infof("Skipping mirroring for %s/%s", endpoints.Namespace, endpoints.Name)
+		return
+	}
+	c.queueEndpoints(obj)
+}
+
+// onEndpointsDelete queues a sync for the relevant Endpoints resource.
+func (c *Controller) onEndpointsDelete(obj interface{}) {
+	endpoints := getEndpointsFromDeleteAction(obj)
+	if endpoints == nil {
+		utilruntime.HandleError(fmt.Errorf("onEndpointsDelete() expected type v1.Endpoints, got %T", obj))
+		return
+	}
+	if !c.shouldMirror(endpoints) {
+		klog.V(5).Infof("Skipping mirroring for %s/%s", endpoints.Namespace, endpoints.Name)
+		return
+	}
+	c.queueEndpoints(obj)
+}
+
+// onEndpointSliceAdd queues a sync for the relevant Endpoints resource for a
+// sync if the EndpointSlice resource version does not match the expected
+// version in the endpointSliceTracker.
+func (c *Controller) onEndpointSliceAdd(obj interface{}) {
+	endpointSlice := obj.(*discovery.EndpointSlice)
+	if endpointSlice == nil {
+		utilruntime.HandleError(fmt.Errorf("onEndpointSliceAdd() expected type discovery.EndpointSlice, got %T", obj))
+		return
+	}
+	if managedByController(endpointSlice) && c.endpointSliceTracker.stale(endpointSlice) {
+		c.queueEndpointsForEndpointSlice(endpointSlice)
+	}
+}
+
+// onEndpointSliceUpdate queues a sync for the relevant Endpoints resource for a
+// sync if the EndpointSlice resource version does not match the expected
+// version in the endpointSliceTracker or the managed-by value of the
+// EndpointSlice has changed from or to this controller.
+func (c *Controller) onEndpointSliceUpdate(prevObj, obj interface{}) {
+	prevEndpointSlice := obj.(*discovery.EndpointSlice)
+	endpointSlice := prevObj.(*discovery.EndpointSlice)
+	if endpointSlice == nil || prevEndpointSlice == nil {
+		utilruntime.HandleError(fmt.Errorf("onEndpointSliceUpdated() expected type discovery.EndpointSlice, got %T, %T", prevObj, obj))
+		return
+	}
+	if managedByChanged(prevEndpointSlice, endpointSlice) || (managedByController(endpointSlice) && c.endpointSliceTracker.stale(endpointSlice)) {
+		c.queueEndpointsForEndpointSlice(endpointSlice)
+	}
+}
+
+// onEndpointSliceDelete queues a sync for the relevant Endpoints resource for a
+// sync if the EndpointSlice resource version does not match the expected
+// version in the endpointSliceTracker.
+func (c *Controller) onEndpointSliceDelete(obj interface{}) {
+	endpointSlice := getEndpointSliceFromDeleteAction(obj)
+	if endpointSlice == nil {
```

Comment:
I thought there is a standard code path where the obj is not of type EndpointSlice, but is some other sentinel value (we've had spurious errors due to this)

We should probably handle that explicitly is that is the case

Reply:
That's handled by `getEndpointSliceFromDeleteAction()`. (That function is identical to EndpointSlice controller implementation).